File: 2018/csharp/day06.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.IO;
using System.Linq;

namespace aoc2018
{
    class Day6
    {
        public static void Solve()
        {            
            var input = File.ReadLines(Path.Combine("..", "input", "day6.input")).ToList();
            var coords = new Dictionary<char, Point>();
            int minx = int.MaxValue, miny = int.MaxValue;
            int maxx = 0, maxy = 0;
            var areas = new Dictionary<char, int>();
            int safePositions = 0;            

            for (int i = 0; i < input.Count; i++)
            {
                var s = input[i].Split(",");
                var p = new Point(int.Parse(s[0]), int.Parse(s[1]));
                coords[(char)('a' + i)] = p;
                minx = Math.Min(int.Parse(s[0]), minx);
                miny = Math.Min(int.Parse(s[1]), miny);
                maxx = Math.Max(int.Parse(s[0]), maxx);
                maxy = Math.Max(int.Parse(s[1]), maxy);
                areas[(char)('a' + i)] = 0;
            }

            for (int y = miny; y <= maxy; y++)
                for (int x = minx; x <= maxx; x++)
                {
                    int minDistance = int.MaxValue;
                    List<char> closestNodes = new List<char>();
                    int sumDistance = 0;
                    foreach (var p in coords.Keys)
                    {
                        int distance = Math.Abs(x - coords[p].X) + Math.Abs(y - coords[p].Y);
                        if (distance < minDistance)
                        {
                            minDistance = distance;
                            closestNodes = new List<char> { p };
                        }
                        else if (distance == minDistance)
                            closestNodes.Add(p);
                        sumDistance += distance;
                    }

                    if (closestNodes.Count == 1)
                        areas[closestNodes.Single()]++;
                    if (sumDistance < 10000)
                        safePositions++;                   
                }
            coords
                .Where((kvp) => kvp.Value.X == minx || kvp.Value.X == maxx || kvp.Value.Y == miny || kvp.Value.Y == maxy)
                .Select(x => x.Key).ToList()
                .ForEach(x => areas.Remove(x));
            Console.WriteLine("day6, part a: {0}", areas.Values.Max());
            Console.WriteLine("day6, part b: {0}", safePositions);
        }
    }
}

File: 2018/python/day06.py
lines = open("../input/day6.input").readlines()
coords = dict()
minx, miny = 1e6, 1e6
maxx, maxy, safePositions = 0, 0, 0
closedAreas = dict()

for i in range(len(lines)):
    s = lines[i].split(",")    
    coords[i] = [int(s[0]), int(s[1])]
    minx = min(int(s[0]), minx)
    miny = min(int(s[1]), miny)
    maxx = max(int(s[0]), maxx)
    maxy = max(int(s[1]), maxy)
    closedAreas[i] = 0

for y in range(miny, maxy + 1):
    for x in range(minx, maxx + 1):
        minDistance, sumDistance = 1e6, 0
        closestNodes= []        
        for p in coords:            
            distance = abs(x - coords[p][0]) + abs(y - coords[p][1])
            if distance < minDistance:
                minDistance = distance
                closestNodes = [ p ]
            elif distance == minDistance:
                closestNodes.append(p)
            sumDistance += distance

        closedAreas[closestNodes[0]]+=1
        if sumDistance < 10000:
            safePositions += 1

infAreas = [k for k, v in coords.items() if v[0] == minx or v[0] == maxx or v[1] == miny or v[1] == maxy]
closedAreas = {k:v for k, v in closedAreas.items() if k not in infAreas}

print ("day 6 part a:", max(closedAreas.values()))
print ("day 6 part b:", safePositions)
